modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/H2SchemaManager.java

```diff
@@ -0,0 +1,418 @@
+import org.h2.index.Index;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * H2 schema manager. Responsible for reflecting to H2 database all schema changes.
+ */
+public class H2SchemaManager implements SchemaChangeListener {
+    /** Connection manager. */
+    private final ConnectionManager connMgr;
+
+    /** Data tables. */
+    private final ConcurrentMap<QueryTable, GridH2Table> dataTables = new ConcurrentHashMap<>();
+
+    /** Kernal context. */
+    private final GridKernalContext ctx;
+
+    /** Indexing. */
+    private final IgniteH2Indexing idx;
+
+    /** H2 index factory. */
+    private final H2IndexFactory idxFactory;
+
+    /** Logger. */
+    private final IgniteLogger log;
+
+    /** Underlying core schema manager. */
+    private volatile SchemaManager schemaMgr;
+
+    /**
+     * Constructor.
+     *
+     * @param ctx Kernal context.
+     * @param idx Indexing.
+     * @param connMgr Connection manager.
+     */
+    public H2SchemaManager(GridKernalContext ctx, IgniteH2Indexing idx, ConnectionManager connMgr) {
+        this.ctx = ctx;
+        this.idx = idx;
+        this.connMgr = connMgr;
+
+        idxFactory = new H2IndexFactory(ctx);
+        log = ctx.log(H2SchemaManager.class);
+    }
+
+    /**
+     * Handle node start.
+     */
+    public void start() throws IgniteCheckedException {
+        schemaMgr = ctx.query().schemaManager();
+
+        ctx.internalSubscriptionProcessor().registerSchemaChangeListener(this);
+
+        // Register predefined system functions.
+        createSqlFunction(QueryUtils.DFLT_SCHEMA, "QUERY_ENGINE", true,
+            H2Utils.class.getName() + ".queryEngine");
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onSchemaCreated(String schema) {
+        try {
+            connMgr.executeSystemStatement("CREATE SCHEMA IF NOT EXISTS " + H2Utils.withQuotes(schema));
+
+            if (log.isDebugEnabled())
+                log.debug("Created H2 schema for index database: " + schema);
+        }
+        catch (IgniteCheckedException e) {
+            throw new IgniteException("Failed to create database schema: " + schema, e);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onSchemaDropped(String schema) {
+        try {
+            connMgr.executeSystemStatement("DROP SCHEMA IF EXISTS " + H2Utils.withQuotes(schema));
+
+            if (log.isDebugEnabled())
+                log.debug("Dropped H2 schema for index database: " + schema);
+        }
+        catch (IgniteCheckedException e) {
+            throw new IgniteException("Failed to drop database schema: " + schema, e);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onSqlTypeCreated(
+        String schemaName,
+        GridQueryTypeDescriptor typeDesc,
+        GridCacheContextInfo<?, ?> cacheInfo
+    ) {
+        H2TableDescriptor tblDesc = new H2TableDescriptor(idx, schemaName, typeDesc, cacheInfo);
```

Comment:
Please add the synchonization between `dataTable` method and table creation.
Without synchonization there is a race: query syccessfully parsed because H2 table is creaed by SQL and dataTables doesn't contains the instance of the `GridH2Table`.

Reply:
As far as I understand there is only method `H2Utils.collectCacheIds` where race between SQL and `dataTable`. I've fixed this method to use `SchemaManager.table()`, which is protected by lock.